src/Catalog/Icons/CatalogLeafDataProcessor.cs

```diff
@@ -0,0 +1,273 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using NuGet.Services.Metadata.Catalog.Helpers;
+using NuGet.Services.Metadata.Catalog.Persistence;
+
+namespace NuGet.Services.Metadata.Catalog.Icons
+{
+    public class CatalogLeafDataProcessor : ICatalogLeafDataProcessor
+    {
+        private const int MaxExternalIconIngestAttempts = 3;
+        private const int MaxBlobStorageCopyAttempts = 3;
+
+        private readonly IAzureStorage _packageStorage;
+        private readonly IIconProcessor _iconProcessor;
+        private readonly IExternalIconContentProvider _externalIconContentProvider;
+        private readonly IIconCopyResultCache _iconCopyResultCache;
+        private readonly ITelemetryService _telemetryService;
+        private readonly ILogger<CatalogLeafDataProcessor> _logger;
+
+        public CatalogLeafDataProcessor(
+            IAzureStorage packageStorage,
+            IIconProcessor iconProcessor,
+            IExternalIconContentProvider externalIconContentProvider,
+            IIconCopyResultCache iconCopyResultCache,
+            ITelemetryService telemetryService,
+            ILogger<CatalogLeafDataProcessor> logger
+            )
+        {
+            _packageStorage = packageStorage ?? throw new ArgumentNullException(nameof(packageStorage));
+            _iconProcessor = iconProcessor ?? throw new ArgumentNullException(nameof(iconProcessor));
+            _externalIconContentProvider = externalIconContentProvider ?? throw new ArgumentNullException(nameof(externalIconContentProvider));
+            _iconCopyResultCache = iconCopyResultCache ?? throw new ArgumentNullException(nameof(iconCopyResultCache));
+            _telemetryService = telemetryService ?? throw new ArgumentNullException(nameof(telemetryService));
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public async Task ProcessPackageDelete(Storage storage, CatalogCommitItem item, CancellationToken cancellationToken)
```

Comment:
Does this method deleted the icon or the entire package? If only the icon maybe the method name can be changed to reflect this?

Reply:
It processes the catalog leaf that indicated that the package was deleted. Renamed.